### ISSUE TITLE
CustomArgumentInfo.previousArgs is empty (should contain PlayerArgument)

### ISSUE THREAD

**Issue by BeefDev** [JorelAli/CommandAPI#298]
**CommandAPI version:** 8.4.1

**Minecraft version:** 1.18.2

**Shading the CommandAPI?** yes

**What I did:**

My plugin stores "profiles" (just an object in memory) for all players. Admins should be able to access profiles by other players, for that purpose I have two arguments. SenderOwnedProfileArgument, which gets the player who is the command sender and gets their profiles, which works perfectly. TargetOwnedProfileArgument, for the admin commands, which gets a previous arg, casts it to a player and follows the same procedure, this breaks.

This is the code for the custom argument which breaks: https://www.toptal.com/developers/hastebin/cubihezazu.java
This is the line where I add the arguments: `.withArguments(new PlayerArgument("target"), new TargetOwnedProfileArgument(this.plugin.getAPI().getPlayerList(), messages, "profile", 0).get())`
This is the command I run: /adminprofiles profile BeefDevClone a

The withArguments method is called on a subcommand (profile) which is later added to adminprofiles as a subcommand. BeefDevClone is supposed to be the player argument (name of my account), and a is the name of one BeefDevClone's profiles.

**What actually happened:**

After I executed the command I get a `Error in executing command adminprofiles profile BeefDevClone a - a<--[HERE]`. Printing inside the function which creates the argument (before line 35 in the code). Shows that the previousArgs array is empty, even though the player argument is before the custom one. This causes the call in line 35 to fail because the array is empty (I assume an ArrayIndexOutOfBoundsException is thrown somewhere.).

Sadly no error logs apart from this.

**What should have happened:**

The PlayerArgument is added to the array, the custom argument gets it and executes normally.

**Comment by BeefDev** [comment]
After some testing, I discovered that a LiteralArgument is placed before other arguments. The loop which attempts to find all previous arguments stops on that one because it matches the node name. That command has a subcommand with the exact same name as the nodename of the argument. In place of that subcommand the LiteralArgument is placed. After reading the source of LiteralArgument, I believe that it is indeed there in place of the subcommand. Since the subcommand is "profile" and the nodename of the argument is "profile". Perhaps the loop thinks the LiteralArgument (which is the first one (index 0)) is the custom one, so it stops the loop there and returns the empty array.

I dont really want to submit a PR since I cant think of an elegant way to solve this or what problems it would cause, but if LiteralArguments are only used there maybe you could exclude them from the loop and start after the last LiteralArgument?

**Comment by willkroboth** [comment]
> The loop which attempts to find all previous arguments stops on that one because it matches the node name.

More specifically, the problem comes from the method `CommandAPIHandler#generatePreviousArguments`
```java
    Object[] generatePreviousArguments(CommandContext<CommandSourceStack> context, Argument<?>[] args, String nodeName)
			throws CommandSyntaxException {
		// Populate Object[], which is our previously filled arguments
		List<Object> previousArguments = new ArrayList<>();

		for (Argument<?> arg : args) {
			if (arg.getNodeName().equals(nodeName)) { // problem here
				break;
			}

			Object result;
			try {
				result = parseArgument(context, arg.getNodeName(), arg, args);
			} catch (IllegalArgumentException e) {
				/*
				 * Redirected commands don't parse previous arguments properly. Simplest way to
				 * determine what we should do is simply set it to null, since there's nothing
				 * else we can do. I thought about letting this simply be an empty array, but
				 * then it's even more annoying to deal with - I wouldn't expect an array of
				 * size n to suddenly, randomly be 0, but I would expect random NPEs because
				 * let's be honest, this is Java we're dealing with.
				 */
				result = null;
			}
			if (arg.isListed()) {
				previousArguments.add(result);
			}
		}
		return previousArguments.toArray();
	}
```
The loop that adds objects to the array returned by `CustomArgumentInfo#previousArgs` breaks once it gets to an argument with the node name equal to the argument for which it is generating the previous arguments. This works as intended if there are no arguments with the same name, which is not necessarily a valid assumption.

The most obvious fix for this issue is to change `arg.getNodeName().equals(nodeName)` to something else that accurately identifies the argument to stop on. I think a good way to do this would be to pass in the entire argument object instead of just the node name and do a `==` check, though I haven't checked that this is compatible with other calls to this method.

Another way to fix this is to avoid calling this method in the first place, at least in this situation. Right now, `CommandAPIHandler#generatePreviousArguments` is being called by `CommandAPIHandler#parseArgument`, which is called by `CommandAPIHandler#argsToObjectArr`.
```java
    Object[] argsToObjectArr(CommandContext<CommandSourceStack> cmdCtx, Argument<?>[] args)
			throws CommandSyntaxException {
		// Array for arguments for executor
		List<Object> argList = new ArrayList<>();

		// Populate array
		for (Argument<?> argument : args) {
			if (argument.isListed()) {
                                // calls parseArgument
				argList.add(parseArgument(cmdCtx, argument.getNodeName(), argument, args));
			}
		}

		return argList.toArray();
	}

    Object parseArgument(CommandContext<CommandSourceStack> cmdCtx, String key, Argument<?> value, Argument<?>[] args)
			throws CommandSyntaxException {
		if (value.isListed()) {
                        // calls generatePreviousArguments
			return value.parseArgument(NMS, cmdCtx, key, generatePreviousArguments(cmdCtx, args, key));
		} else {
			return null;
		}
	}

    Object[] generatePreviousArguments(CommandContext<CommandSourceStack> context, Argument<?>[] args, String nodeName)
			throws CommandSyntaxException {
		// Populate Object[], which is our previously filled arguments
		List<Object> previousArguments = new ArrayList<>();

		for (Argument<?> arg : args) {
			if (arg.getNodeName().equals(nodeName)) { // problem here
				break;
			}

			Object result;
			try {
                                // calls parseArgument
				result = parseArgument(context, arg.getNodeName(), arg, args);
			} catch (IllegalArgumentException e) {
				result = null;
			}
			if (arg.isListed()) {
				previousArguments.add(result);
			}
		}
		return previousArguments.toArray();
	}
```
For each arg that needs to be parsed by `argsToObjectArray`, `parseArgument` is called, which calls `generatePreviousArguments`, which tries to parse all of the previous arguments. However, `argsToObjectArray` has already parsed these arguments (putting them into `argList`), so the call to `generatePreviousArguments` seems redundant. Instead of parsing these arguments again, `parseArguments` could just request `argList` from `argsToObjectArray`. Again, I haven't checked if this solution has any unintended consequences, but I'm planning to look into this.

**Comment by JorelAli** [comment]
@willkroboth 
> The loop that adds objects to the array returned by CustomArgumentInfo#previousArgs breaks once it gets to an argument with the node name equal to the argument for which it is generating the previous arguments. **This works as intended if there are no arguments with the same name, which is not necessarily a valid assumption.**

Well yes, but technically no. Brigadier requires _required arguments_ to have different names (i.e. it applies to all arguments except literal arguments). Arguments with the same name do conflict internally as well in Brigadier (implemented in [`CommandContext#getArgument`](https://github.com/Mojang/brigadier/blob/master/src/main/java/com/mojang/brigadier/context/CommandContext.java#L82), which uses a map of argument names to parsed arguments). I'm fairly certain simply omitting the case of literal arguments would be sufficient.

Everything else you said about calling the generatePreviousArguments multiple times sounds pretty plausible however! I'm all for any performance improvements :)

**Comment by willkroboth** [comment]
I created a pull request that addresses this issue via the method I mentioned second. This removes the call to `generatePreviousArguments` from `parseArgument`. It looks like this means `generatePreviousArguments` is only called by one other method, `CommandAPIHandler#toSuggestions`:
```java
SuggestionProvider<CommandSourceStack> toSuggestions(String nodeName, Argument<?>[] args,
		boolean overrideSuggestions) {
	return (CommandContext<CommandSourceStack> context, SuggestionsBuilder builder) -> {
		SuggestionInfo suggestionInfo = new SuggestionInfo(NMS.getCommandSenderFromCSS(context.getSource()),
                                // right here
				generatePreviousArguments(context, args, nodeName), builder.getInput(), builder.getRemaining());
		Optional<ArgumentSuggestions> suggestionsToAddOrOverride = overrideSuggestions
				? getArgument(args, nodeName).getOverriddenSuggestions()
				: getArgument(args, nodeName).getIncludedSuggestions();
		return suggestionsToAddOrOverride.orElse(ArgumentSuggestions.empty()).suggest(suggestionInfo, builder);
	};
}
```
I haven't figured out what this `toSuggestions` method is used for, but this report reveals that `generatePreviousArguments` can provide the wrong result when arguments have the same name, so I wonder if there is a hidden issue here.

> Brigadier requires _required arguments_ to have different names

Maybe something in Brigadier prevents this from being a problem.

**Comment by JorelAli** [comment]
If I recall correctly, Brigadier doesn't prevent you from having multiple arguments with the same name, but the general behaviour of a command can be undetermined. I can't remember (off the top of my head) if the CommandAPI safeguards against this (it should). I'll stare at this in a bit more detail later (probably some time after 18th July, or sooner if I have any spare time) and will investigate how literal arguments could/should be handled.

**Comment by JorelAli** [comment]
This should now be fixed in a0b059a09fb2f7e0b9ffbe209d55625c149a299b on the `dev/dev` branch for release 8.5.1. Since the original example code isn't available anymore (hastebin), this original issue cannot be properly checked so I'll be going off of "it probably works now".

**Comment by JorelAli** [comment]
Fixed in 8.5.1.
